File: sdks/java/basys.sdk/src/org/eclipse/basyx/aas/metamodel/hashmap/aas/qualifier/Referable.java
package org.eclipse.basyx.aas.metamodel.hashmap.aas.qualifier;

import java.util.HashMap;

import org.eclipse.basyx.aas.api.metamodel.aas.qualifier.IReferable;
import org.eclipse.basyx.aas.api.metamodel.aas.reference.IReference;
import org.eclipse.basyx.aas.metamodel.facades.ReferableFacade;

/**
 * Referable class
 * 
 * @author kuhn, schnicke
 *
 */
public class Referable extends HashMap<String, Object> implements IReferable {

	/**
	 * Version of serialized instances
	 */
	private static final long serialVersionUID = 1L;
	
	public static final String IDSHORT="idShort";
	
	public static final String CATEGORY="category";
	
	public static final String DESCRIPTION="description";
	
	public static final String PARENT="parent";

	/**
	 * Constructor
	 */
	public Referable() {
		// Identifies an element within its name space (String)
		put(IDSHORT, "");
		// Coded value that gives further meta information w.r.t. to the type of the
		// element. It affects the
		// expected existence of attributes and the applicability of constraints.
		// (String)
		put(CATEGORY, "");
		// Description or comments on the element (String)
		put(DESCRIPTION, "");
		// Reference to the parent of this element (Referable)
		put(PARENT, null);
	}

	/**
	 * Constructor with idShort, category and description
	 * 
	 * @param idShort
	 *            will be matched case insensitive; may only feature letters,
	 *            digits, underscores and start with a letter
	 * @param category
	 * @param description
	 */
	public Referable(String idShort, String category, String description) {
		// Identifies an element within its name space (String)
		put(IDSHORT, idShort);
		// Coded value that gives further meta information w.r.t. to the type of the
		// element. It affects the
		// expected existence of attributes and the applicability of constraints.
		// (String)
		put(CATEGORY, category);
		// Description or comments on the element (String)
		put(DESCRIPTION, description);
		// Reference to the parent of this element (Referable)
		put(PARENT, null);
	}

	@Override
	public String getIdshort() {
	return new ReferableFacade(this).getIdshort();
	}

	@Override
	public String getCategory() {
		return new ReferableFacade(this).getCategory();
	}

	@Override
	public String getDescription() {
		return new ReferableFacade(this).getDescription();
	}

	@Override
	public IReference  getParent() {
		return new ReferableFacade(this).getParent();
	}

	@Override
	public void setIdshort(String idShort) {
		new ReferableFacade(this).setIdshort(idShort);
		
	}

	@Override
	public void setCategory(String category) {
		new ReferableFacade(this).setCategory(category);
		
	}

	@Override
	public void setDescription(String description) {
		new ReferableFacade(this).setDescription(description);
		
	}

	@Override
	public void setParent(IReference  obj) {
		new ReferableFacade(this).setParent(obj);
		
	}


}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map/qualifier/Referable.cpp
/*
 * Referable.cpp
 *
 *      Author: wendel
 */

#include "Referable.h"
#include "submodel/map/reference/Reference.h"

namespace basyx {
namespace submodel {

Referable::Referable()
  : vab::ElementMap{}
{
}

basyx::submodel::Referable::Referable(basyx::object & obj)
	: vab::ElementMap{ obj }
{
	this->map.insertKey(Path::IdShort, "");
	this->map.insertKey(Path::Category, "");
	this->map.insertKey(Path::Description, Description{}.getMap());
	this->map.insertKey(Path::Parent, basyx::object::make_null());
}

Referable::Referable(const std::string & idShort, const std::string & category, const Description & description) :
  vab::ElementMap{}
{
	this->map.insertKey(Path::IdShort, idShort);
	this->map.insertKey(Path::Category, category);
	this->map.insertKey(Path::Description, description.getMap());
	this->map.insertKey(Path::Parent, basyx::object::make_null());
}

Referable::Referable(const IReferable & other) :
  vab::ElementMap{}
{
  this->setIdShort(other.getIdShort());
  this->setCategory(other.getCategory());
  this->setDescription(other.getDescription());
  this->setParent(other.getParent());
}

Referable::Referable(const std::shared_ptr<IReferable>& other)

{
  this->setIdShort(other->getIdShort());
  this->setCategory(other->getCategory());
  this->setDescription(other->getDescription());
  this->setParent(other->getParent());
}

std::string Referable::getIdShort() const
{
	return this->map.getProperty(Path::IdShort).GetStringContent();
}

std::string Referable::getCategory() const
{
	return this->map.getProperty(Path::Category).GetStringContent();
}

Description Referable::getDescription() const
{
	return Description{ this->map.getProperty(Path::Description) };
}

std::shared_ptr<IReference> Referable::getParent() const
{
	return std::make_shared<Reference>(this->map.getProperty(Path::Parent));
}

void Referable::setIdShort(const std::string & shortID)
{
	this->map.insertKey(Path::IdShort, shortID, true);
}

void Referable::setCategory(const std::string & category)
{
	this->map.insertKey(Path::Category, category, true);
}

void Referable::setDescription(const Description & description)
{
	this->map.insertKey(Path::Description, description.getMap(), true);
}

void Referable::setParent(const std::shared_ptr<IReference> & parentReference)
{
  this->insertMapElement(Path::Parent, Reference{parentReference});
}

}
}
